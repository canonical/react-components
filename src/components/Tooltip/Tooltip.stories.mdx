import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";
import Button from "../Button";
import Tooltip, { position, variant } from "./Tooltip";

<Meta
  title="Tooltip"
  component={Tooltip}
  argTypes={{
    position: {
      control: {
        type: "radio",
      },
      options: Object.values(position),
    },
    variant: {
      control: {
        type: "radio",
      },
      options: Object.values(variant),
    },
  }}
/>

export const Template = (args) => {
  return (
    <div style={{ display: "flex", justifyContent: "center", padding: "3rem" }}>
      <Tooltip {...args}>
        <Button className="u-no-margin--bottom">Hover over me!</Button>
      </Tooltip>
    </div>
  );
};

### Tooltip

This is the [React](https://reactjs.org/) component for Vanilla [Tooltips](https://vanillaframework.io/docs/patterns/tooltips).

Tooltips are text labels that appear when the user hovers over, focuses on, or touches an element on the screen.

They can be used to provide information about concepts/terms/actions that are not self-explanatory or well known.

An alternative use of tooltips is to provide information on a disabled actionable element, e.g. for disabled buttons, providing information about why they are disabled.

### Props

<ArgsTable of={Tooltip} />

### Regular

<Canvas>
  <Story
    name="Regular"
    args={{
      variant: "regular",
      position: "right",
      message: <>Tooltip <a href="#todo" target="_blank">Link1</a></>
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Detached (default)

<Canvas>
  <Story
    name="Detached (default)"
    args={{
      variant: "detached",
      message: <>Tooltip <a href="#todo" target="_blank">Link1</a></>
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Follow mouse

<Canvas>
  <Story
    name="Follow mouse"
    args={{
      variant: "detached",
      followMouse: true,
      message: "Tooltip message to display.",
      position: "top-center",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Targeting elements

The tooltip elements are displayed inside a [Portal](https://reactjs.org/docs/portals.html) so that the tooltip is not constrained by its current position in the DOM.
Sometimes however, you need to target the tooltip elements. This can be achieved by passing class names to the tooltip elements. You can also pass classes to the elements that trigger and position the tooltip.

<Canvas>
  <Story
    name="Targeting elements"
    args={{
      variant: "detached",
      className: "custom-wrapping-class",
      positionElementClassName: "custom-position-class",
      tooltipClassName: "custom-tooltip-class",
      message: "Tooltip message to display.",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>
