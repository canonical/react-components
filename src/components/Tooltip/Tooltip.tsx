import classNames from "classnames";
import React, { useCallback, useEffect, useId, useRef, useState } from "react";
import type { MouseEventHandler, ReactNode } from "react";
import usePortal from "react-useportal";

import { useWindowFitment, useListener } from "hooks";
import type { WindowFitment } from "hooks";

import type { ClassName, ValueOf } from "types";

export type CSSPosition =
  | "static"
  | "absolute"
  | "fixed"
  | "relative"
  | "sticky"
  | "initial"
  | "inherit";

export type PositionStyle = {
  left: number;
  pointerEvents?: string;
  position: CSSPosition;
  top: number;
};

export const position = {
  btmCenter: "btm-center",
  btmLeft: "btm-left",
  btmRight: "btm-right",
  left: "left",
  right: "right",
  topCenter: "top-center",
  topLeft: "top-left",
  topRight: "top-right",
} as const;

export type Position = ValueOf<typeof position>;

export type Props = {
  /**
   * Whether the tooltip should adjust to fit in the screen.
   */
  autoAdjust?: boolean;
  /**
   * Element on which to apply the tooltip.
   */
  children: ReactNode;
  /**
   * An optional class to apply to the wrapping element.
   */
  className?: ClassName;
  /**
   * Whether the tooltip should follow the mouse.
   */
  followMouse?: boolean;
  /**
   * Message to display when the element is hovered.
   */
  message?: ReactNode;
  /**
   * Position of the tooltip relative to the element.
   */
  position?: Position;
  /**
   * An optional class to apply to the element that wraps the children.
   */
  positionElementClassName?: string;
  /**
   * An optional class to apply to the tooltip message element.
   */
  tooltipClassName?: string;
  /**
   * The z-index value of the tooltip message element.
   */
  zIndex?: number;
  /**
   * Delay in ms after which Tooltip will appear (defaults to 350ms).
   */
  delay?: number;
};

const getPositionStyle = (
  pos: Position,
  wrapperNode: HTMLElement
): PositionStyle => {
  if (!wrapperNode) {
    return null;
  }

  const dimensions = wrapperNode.getBoundingClientRect();
  const { x, y, height, width } = dimensions;
  let left = x + window.scrollX || 0;
  let top = y + window.scrollY || 0;

  switch (pos) {
    case "btm-center":
      left += width / 2;
      top += height;
      break;
    case "btm-left":
      top += height;
      break;
    case "btm-right":
      left += width;
      top += height;
      break;
    case "left":
      top += height / 2;
      break;
    case "right":
      left += width;
      top += height / 2;
      break;
    case "top-center":
      left += width / 2;
      break;
    case "top-left":
      break;
    case "top-right":
      left += width;
      break;
    default:
      break;
  }
  return { position: "absolute", left, top };
};

export const adjustForWindow = (
  position: Position,
  fitsWindow: WindowFitment
): Position => {
  let newPosition: string = position;
  if (!fitsWindow.fromLeft.fitsLeft && newPosition === "left") {
    newPosition = "top-right";
  }
  if (!fitsWindow.fromRight.fitsRight && newPosition === "right") {
    newPosition = "top-left";
  }
  if (!fitsWindow.fromRight.fitsLeft && newPosition.endsWith("-right")) {
    newPosition = newPosition.replace("right", "left");
  }
  if (!fitsWindow.fromLeft.fitsRight && newPosition.endsWith("-left")) {
    newPosition = newPosition.replace("left", "right");
  }
  if (!fitsWindow.fromTop.fitsAbove && newPosition.startsWith("top")) {
    newPosition = newPosition.replace("top", "btm");
  }
  if (!fitsWindow.fromBottom.fitsBelow && newPosition.startsWith("btm")) {
    newPosition = newPosition.replace("btm", "top");
  }
  if (
    !fitsWindow.fromLeft.fitsRight &&
    !fitsWindow.fromRight.fitsLeft &&
    (newPosition.endsWith("-left") || newPosition.endsWith("-right"))
  ) {
    newPosition = newPosition
      .replace("left", "center")
      .replace("right", "center");
  }
  if (
    newPosition.endsWith("center") &&
    (fitsWindow.fromCenter.fitsCentered.fitsRight ||
      fitsWindow.fromCenter.fitsCentered.fitsLeft)
  ) {
    if (!fitsWindow.fromCenter.fitsCentered.fitsRight) {
      newPosition = newPosition.replace("center", "right");
    }
    if (!fitsWindow.fromCenter.fitsCentered.fitsLeft) {
      newPosition = newPosition.replace("center", "left");
    }
  }
  return newPosition as Position;
};

/**
 * This is the [React](https://reactjs.org/) component for Vanilla [Tooltips](https://vanillaframework.io/docs/patterns/tooltips).
 *
 * Tooltips are text labels that appear when the user hovers over, focuses on, or touches an element on the screen.
 *
 * They can be used to provide information about concepts/terms/actions that are not self-explanatory or well known.
 *
 * An alternative use of tooltips is to provide information on a disabled actionable element, e.g. for disabled buttons, providing information about why they are disabled.
 */
const Tooltip = ({
  autoAdjust = true,
  children,
  className,
  followMouse = false,
  message,
  position = "top-left",
  positionElementClassName,
  tooltipClassName,
  zIndex,
  delay = 350,
}: Props): JSX.Element => {
  const wrapperRef = useRef<HTMLElement>(null);
  const messageRef = useRef<HTMLElement>(null);
  const [adjustedPosition, setAdjustedPosition] = useState(position);
  const [positionStyle, setPositionStyle] = useState<PositionStyle>({
    position: "absolute",
    // Initially position the tooltip of the screen in case it gets shown
    // before setting the position.
    left: -9999999,
    top: -9999999,
  });
  const { openPortal, closePortal, isOpen, Portal } = usePortal({
    programmaticallyOpen: true,
  });
  const tooltipId = useId();
  const [timer, setTimer] = useState<ReturnType<typeof setTimeout> | null>(
    null
  );
  const cancelableClosePortal = useCallback(() => {
    clearTimeout(timer);
    closePortal();
  }, [timer, closePortal]);

  useEffect(() => {
    if (isOpen && !followMouse && wrapperRef.current) {
      // Position the tooltip when it becomes visible.
      setPositionStyle(getPositionStyle(adjustedPosition, wrapperRef.current));
    }
  }, [adjustedPosition, isOpen, followMouse]);

  const mouseHandler = useCallback((evt: MouseEvent) => {
    // Set the position of the tooltip next to the mouse.
    setPositionStyle({
      // Don't allow the tooltip to block the mouse events.
      pointerEvents: "none",
      position: "absolute",
      left: evt.pageX,
      top: evt.pageY,
    });
  }, []);

  const onUpdateWindowFitment = useCallback(
    (fitsWindow) => {
      setAdjustedPosition(adjustForWindow(position, fitsWindow));
    },
    [setAdjustedPosition, position]
  );

  // Handle mouse events.
  useListener(
    wrapperRef.current,
    mouseHandler,
    "mousemove",
    true,
    followMouse && isOpen
  );

  // Handle adjusting the position of the tooltip so that it remains on screen.
  useWindowFitment(
    messageRef.current,
    wrapperRef.current,
    onUpdateWindowFitment,
    20,
    isOpen,
    autoAdjust && followMouse
  );

  const handleKeyPress = useCallback(
    (event: KeyboardEvent) => {
      if (event.key === "Escape") {
        cancelableClosePortal();
      }
    },
    [cancelableClosePortal]
  );

  useEffect(() => {
    window.addEventListener("keypress", handleKeyPress);
    return () => {
      window.removeEventListener("keypress", handleKeyPress);
    };
  }, [handleKeyPress]);

  const handleBlur = (e) => {
    // do not close if the focus is within the tooltip wrapper
    if (wrapperRef?.current?.contains(document.activeElement)) {
      return;
    }

    if (
      e.relatedTarget
        ? !messageRef.current?.contains(e.relatedTarget) &&
          !wrapperRef.current?.contains(e.relatedTarget)
        : e.target !== messageRef.current
    ) {
      cancelableClosePortal();
    }
  };

  const handleClick = (e) => {
    // ignore clicks within the tooltip message
    if (messageRef.current?.contains(e.target)) {
      return;
    }
    e.target.focus();
    openPortal(e);
  };

  const delayedOpenPortal: MouseEventHandler = useCallback(() => {
    if (isOpen) {
      return;
    }
    if (timer) {
      clearTimeout(timer);
    }
    const timeout = setTimeout(() => openPortal(), delay);
    setTimer(timeout);
  }, [delay, openPortal, timer, isOpen]);

  return (
    <>
      {message ? (
        <span
          className={className}
          onBlur={handleBlur}
          onClick={handleClick}
          onFocus={openPortal}
          onMouseOut={handleBlur}
          onMouseOver={delayedOpenPortal}
        >
          <span
            className={positionElementClassName}
            ref={wrapperRef}
            style={{ display: "inline-block" }}
          >
            {React.Children.map(children, (child) =>
              child && React.isValidElement(child)
                ? React.cloneElement(child, {
                    "aria-describedby": isOpen ? tooltipId : undefined,
                  } as React.HTMLAttributes<HTMLElement>)
                : child
            )}
          </span>
          {isOpen ? (
            <Portal>
              <span
                className={classNames(
                  `p-tooltip--${adjustedPosition}`,
                  "is-detached",
                  tooltipClassName
                )}
                data-testid="tooltip-portal"
                style={positionStyle as React.CSSProperties}
              >
                <span
                  role="tooltip"
                  className="p-tooltip__message"
                  onClick={(event) => {
                    // Prevent clicks inside the message from bubbling to parent
                    // click handlers.
                    event.stopPropagation();
                  }}
                  ref={messageRef}
                  id={tooltipId}
                  style={{ zIndex: zIndex }}
                >
                  {message}
                </span>
              </span>
            </Portal>
          ) : null}
        </span>
      ) : (
        <span className={className}>{children}</span>
      )}
    </>
  );
};

export default Tooltip;
