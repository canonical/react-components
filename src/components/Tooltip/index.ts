export { default, position, variant } from "./Tooltip";
export type { Props as TooltipProps } from "./Tooltip";
