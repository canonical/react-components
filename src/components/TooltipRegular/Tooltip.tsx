import classNames from "classnames";
import React, { useRef } from "react";
import type { ReactNode, ReactElement } from "react";

import { useId } from "hooks";
import type { ClassName } from "types";

export type CSSPosition =
  | "static"
  | "absolute"
  | "fixed"
  | "relative"
  | "sticky"
  | "initial"
  | "inherit";

export type Position =
  | "btm-center"
  | "btm-left"
  | "btm-right"
  | "left"
  | "right"
  | "top-center"
  | "top-left"
  | "top-right";

export type Props = {
  /**
   * Element to attach the tooltip to.
   */
  children: ReactElement;
  /**
   * An optional class to apply to the wrapping element.
   */
  className?: ClassName;
  /**
   * Message to display when the element is hovered.
   */
  message?: ReactNode;
  /**
   * Position of the tooltip relative to the element.
   */
  position?: Position;
};

const Tooltip = ({
  children,
  className,
  message,
  position = "top-left",
}: Props): JSX.Element => {
  const id = useId();
  const tooltipRef = useRef<HTMLElement>(null);

  return (
    <>
      {message ? (
        <>
          {React.Children.map(children, (child) =>
            React.cloneElement(child, {
              "aria-describedby": id,
              className: classNames("p-tooltip", child?.props?.className, {
                [`p-tooltip--${position}`]: !!position,
              }),
              children: (
                <>
                  {child?.props?.children}
                  <span
                    ref={tooltipRef}
                    id={id}
                    role="tooltip"
                    className={classNames("p-tooltip__message", className)}
                  >
                    {message}
                  </span>
                </>
              ),
            })
          )}
        </>
      ) : (
        children
      )}
    </>
  );
};

export default Tooltip;
