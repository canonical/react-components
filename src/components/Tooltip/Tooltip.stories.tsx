import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import Button from "../Button";
import Tooltip, { position } from "./Tooltip";

const Template = (args) => {
  return (
    <div style={{ display: "flex", justifyContent: "center", padding: "3rem" }}>
      <Tooltip {...args}>
        <Button className="u-no-margin--bottom">Hover over me!</Button>
      </Tooltip>
    </div>
  );
};

const meta: Meta<typeof Tooltip> = {
  component: Tooltip,
  render: Template,
  tags: ["autodocs"],

  argTypes: {
    position: {
      control: {
        type: "radio",
      },

      options: Object.values(position),
    },
  },
};

export default meta;

type Story = StoryObj<typeof Tooltip>;

export const Default: Story = {
  render: Template.bind({}),
  name: "Default",

  args: {
    message: (
      <>
        Tooltip{" "}
        <a href="#todo" target="_blank">
          Link1
        </a>
      </>
    ),
  },
};

export const FollowMouse: Story = {
  render: Template.bind({}),
  name: "Follow mouse",

  args: {
    followMouse: true,
    message: "Tooltip message to display.",
    position: "top-center",
  },
};

/**
 * The tooltip elements are displayed inside a [Portal](https://reactjs.org/docs/portals.html) so that the tooltip is not constrained by its current position in the DOM.
 * Sometimes however, you need to target the tooltip elements. This can be achieved by passing class names to the tooltip elements. You can also pass classes to the elements that trigger and position the tooltip.
 */
export const TargetingElements: Story = {
  render: Template.bind({}),
  name: "Targeting elements",

  args: {
    className: "custom-wrapping-class",
    positionElementClassName: "custom-position-class",
    tooltipClassName: "custom-tooltip-class",
    message: "Tooltip message to display.",
  },
};
