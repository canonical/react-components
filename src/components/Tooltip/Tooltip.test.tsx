import { mount, shallow } from "enzyme";
import merge from "deepmerge";
import React from "react";

import Tooltip, { adjustForWindow } from "./Tooltip";

describe("<Tooltip />", () => {
  const body = document.querySelector("body");
  const app = document.createElement("div");
  app.setAttribute("id", "app");
  body.appendChild(app);

  // snapshot tests
  it("renders and matches the snapshot", () => {
    const component = shallow(<Tooltip message="text">Child</Tooltip>);
    expect(component).toMatchSnapshot();
  });

  // unit tests
  it("does not show tooltip message by default", () => {
    const component = mount(<Tooltip message="text">Child</Tooltip>);
    expect(component.exists("[data-testid='tooltip-portal']")).toEqual(false);
  });

  it("renders tooltip message when focused", () => {
    const component = mount(<Tooltip message="text">Child</Tooltip>);
    component.simulate("focus");
    expect(component.find("[data-testid='tooltip-portal']").text()).toEqual(
      "text"
    );
  });

  it("can display elements inside the message", () => {
    const component = mount(
      <Tooltip message={<strong>message</strong>}>Child</Tooltip>
    );
    component.simulate("focus");
    expect(
      component.find("[data-testid='tooltip-portal'] strong").exists()
    ).toBe(true);
  });

  it("gives the correct class name to the tooltip", () => {
    const component = mount(
      <Tooltip message="text" position="right">
        Child
      </Tooltip>
    );
    component.simulate("focus");
    expect(component.exists(".p-tooltip--right")).toEqual(true);
  });

  it("updates the tooltip to fit on the screen", () => {
    const wrapper = mount(
      <Tooltip
        message="text that is too long to fit on the screen"
        position="right"
      >
        Child
      </Tooltip>
    );
    global.innerWidth = 20;
    wrapper.simulate("mouseover");
    expect(
      wrapper.find("[data-testid='tooltip-portal']").prop("className")
    ).toBe("p-tooltip--btm-left is-detached");
  });

  describe("adjustForWindow", () => {
    const generateFits = (overrides = {}) => {
      const fits = {
        fromTop: { fitsAbove: true, fitsBelow: true },
        fromBottom: { fitsAbove: true, fitsBelow: true },
        fromLeft: { fitsLeft: true, fitsRight: true },
        fromRight: { fitsLeft: true, fitsRight: true },
        fromCenter: {
          fitsLeft: true,
          fitsRight: true,
          fitsAbove: true,
          fitsBelow: true,
          fitsCentered: {
            fitsLeft: true,
            fitsRight: true,
            fitsAbove: true,
            fitsBelow: true,
          },
        },
      };
      return merge(fits, overrides);
    };

    it("doesn't change if it fits", () => {
      expect(adjustForWindow("left", generateFits())).toBe("left");
      expect(adjustForWindow("btm-left", generateFits())).toBe("btm-left");
      expect(adjustForWindow("btm-center", generateFits())).toBe("btm-center");
      expect(adjustForWindow("btm-right", generateFits())).toBe("btm-right");
      expect(adjustForWindow("right", generateFits())).toBe("right");
      expect(adjustForWindow("top-right", generateFits())).toBe("top-right");
      expect(adjustForWindow("top-center", generateFits())).toBe("top-center");
      expect(adjustForWindow("top-left", generateFits())).toBe("top-left");
    });

    it("handles left that doesn't fit left", () => {
      expect(
        adjustForWindow(
          "left",
          generateFits({
            fromLeft: { fitsLeft: false },
          })
        )
      ).toBe("top-right");
    });

    it("handles left that doesn't fit left and top", () => {
      expect(
        adjustForWindow(
          "left",
          generateFits({
            fromTop: { fitsAbove: false },
            fromLeft: { fitsLeft: false },
          })
        )
      ).toBe("btm-right");
    });

    it("handles btm-left that doesn't fit right from right edge", () => {
      expect(
        adjustForWindow(
          "btm-left",
          generateFits({
            fromRight: { fitsRight: false },
          })
        )
      ).toBe("btm-left");
    });

    it("handles btm-left that doesn't fit right from left edge", () => {
      expect(
        adjustForWindow(
          "btm-left",
          generateFits({
            fromLeft: { fitsRight: false },
            fromRight: { fitsRight: false },
          })
        )
      ).toBe("btm-right");
    });

    it("handles btm-left that doesn't fit bottom", () => {
      expect(
        adjustForWindow(
          "btm-left",
          generateFits({
            fromBottom: { fitsBelow: false },
          })
        )
      ).toBe("top-left");
    });

    it("handles btm-left that doesn't fit right and bottom", () => {
      expect(
        adjustForWindow(
          "btm-left",
          generateFits({
            fromBottom: { fitsBelow: false },
            fromLeft: { fitsRight: false },
          })
        )
      ).toBe("top-right");
    });

    it("handles btm-right that doesn't fit left from left edge", () => {
      expect(
        adjustForWindow(
          "btm-right",
          generateFits({
            fromLeft: { fitsLeft: false },
          })
        )
      ).toBe("btm-right");
    });

    it("handles btm-right that doesn't fit left from left edge", () => {
      expect(
        adjustForWindow(
          "btm-right",
          generateFits({
            fromLeft: { fitsLeft: false },
            fromRight: { fitsLeft: false },
          })
        )
      ).toBe("btm-left");
    });

    it("handles btm-right that doesn't fit bottom", () => {
      expect(
        adjustForWindow(
          "btm-right",
          generateFits({
            fromBottom: { fitsBelow: false },
          })
        )
      ).toBe("top-right");
    });

    it("handles btm-right that doesn't fit left and bottom", () => {
      expect(
        adjustForWindow(
          "btm-right",
          generateFits({
            fromBottom: { fitsBelow: false },
            fromRight: { fitsLeft: false },
          })
        )
      ).toBe("top-left");
    });

    it("handles right that doesn't fit right", () => {
      expect(
        adjustForWindow(
          "right",
          generateFits({
            fromRight: { fitsRight: false },
          })
        )
      ).toBe("top-left");
    });

    it("handles right that doesn't fit right and top", () => {
      expect(
        adjustForWindow(
          "right",
          generateFits({
            fromTop: { fitsAbove: false },
            fromRight: { fitsRight: false },
          })
        )
      ).toBe("btm-left");
    });

    it("handles top-left that doesn't fit right from right edge", () => {
      expect(
        adjustForWindow(
          "top-left",
          generateFits({
            fromRight: { fitsRight: false },
          })
        )
      ).toBe("top-left");
    });

    it("handles top-left that doesn't fit right from left edge", () => {
      expect(
        adjustForWindow(
          "top-left",
          generateFits({
            fromLeft: { fitsRight: false },
            fromRight: { fitsRight: false },
          })
        )
      ).toBe("top-right");
    });

    it("handles top-left that doesn't fit top", () => {
      expect(
        adjustForWindow(
          "top-left",
          generateFits({
            fromTop: { fitsAbove: false },
          })
        )
      ).toBe("btm-left");
    });

    it("handles top-left that doesn't fit right and top", () => {
      expect(
        adjustForWindow(
          "top-left",
          generateFits({
            fromTop: { fitsAbove: false },
            fromLeft: { fitsRight: false },
          })
        )
      ).toBe("btm-right");
    });

    it("handles top-right that doesn't fit left from left edge", () => {
      expect(
        adjustForWindow(
          "top-right",
          generateFits({
            fromLeft: { fitsLeft: false },
          })
        )
      ).toBe("top-right");
    });

    it("handles top-right that doesn't fit left from right edge", () => {
      expect(
        adjustForWindow(
          "top-right",
          generateFits({
            fromLeft: { fitsLeft: false },
            fromRight: { fitsLeft: false },
          })
        )
      ).toBe("top-left");
    });

    it("handles top-right that doesn't fit top", () => {
      expect(
        adjustForWindow(
          "top-right",
          generateFits({
            fromTop: { fitsAbove: false },
          })
        )
      ).toBe("btm-right");
    });

    it("handles top-right that doesn't fit left and top", () => {
      expect(
        adjustForWindow(
          "top-right",
          generateFits({
            fromTop: { fitsAbove: false },
            fromRight: { fitsLeft: false },
          })
        )
      ).toBe("btm-left");
    });

    it("handles top-right that doesn't fit left or right", () => {
      expect(
        adjustForWindow(
          "top-right",
          generateFits({
            fromLeft: { fitsLeft: false, fitsRight: false },
            fromRight: { fitsLeft: false, fitsRight: false },
          })
        )
      ).toBe("top-center");
    });

    it("handles top-left that doesn't fit left or right", () => {
      expect(
        adjustForWindow(
          "top-left",
          generateFits({
            fromLeft: { fitsLeft: false, fitsRight: false },
            fromRight: { fitsLeft: false, fitsRight: false },
          })
        )
      ).toBe("top-center");
    });

    it("handles top-center that doesn't fit left or right", () => {
      expect(
        adjustForWindow(
          "top-center",
          generateFits({
            fromCenter: { fitsCentered: { fitsLeft: false, fitsRight: false } },
          })
        )
      ).toBe("top-center");
      // top-center
    });

    it("handles top-center that doesn't fit left", () => {
      expect(
        adjustForWindow(
          "top-center",
          generateFits({
            fromCenter: { fitsCentered: { fitsLeft: false } },
          })
        )
      ).toBe("top-left");
    });

    it("handles top-center that doesn't fit right", () => {
      expect(
        adjustForWindow(
          "top-center",
          generateFits({
            fromCenter: { fitsCentered: { fitsRight: false } },
          })
        )
      ).toBe("top-right");
    });

    it("handles btm-right that doesn't fit left or right", () => {
      expect(
        adjustForWindow(
          "btm-right",
          generateFits({
            fromLeft: { fitsLeft: false, fitsRight: false },
            fromRight: { fitsLeft: false, fitsRight: false },
          })
        )
      ).toBe("btm-center");
    });

    it("handles btm-left that doesn't fit left or right", () => {
      expect(
        adjustForWindow(
          "btm-left",
          generateFits({
            fromLeft: { fitsLeft: false, fitsRight: false },
            fromRight: { fitsLeft: false, fitsRight: false },
          })
        )
      ).toBe("btm-center");
    });

    it("handles btm-center that doesn't fit left or right", () => {
      expect(
        adjustForWindow(
          "btm-center",
          generateFits({
            fromCenter: { fitsCentered: { fitsLeft: false, fitsRight: false } },
          })
        )
      ).toBe("btm-center");
    });

    it("handles btm-center that doesn't fit left", () => {
      expect(
        adjustForWindow(
          "btm-center",
          generateFits({
            fromCenter: { fitsCentered: { fitsLeft: false } },
          })
        )
      ).toBe("btm-left");
    });

    it("handles btm-center that doesn't fit right", () => {
      expect(
        adjustForWindow(
          "btm-center",
          generateFits({
            fromCenter: { fitsCentered: { fitsRight: false } },
          })
        )
      ).toBe("btm-right");
    });
  });
});
