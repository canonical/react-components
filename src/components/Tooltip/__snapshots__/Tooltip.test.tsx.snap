// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> renders and matches the snapshot 1`] = `
<Tooltip
  message="text"
>
  <TooltipDetached
    message="text"
  >
    <span
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <span
        style={
          Object {
            "display": "inline-block",
          }
        }
      >
        Child
      </span>
    </span>
  </TooltipDetached>
</Tooltip>
`;
